File: Javascript/DCP08_count_unival.js
class Node
{
  constructor(data, left = null, right = null)
  {
    this.data = data;
    this.left = left;
    this.right = right;
  }

  inorder_traversal(node, tree)
  {
    if(node.left != null)
    {
      this.inorder_traversal(node.left, tree);
    }

    tree.push(node);

    if(node.right != null)
    {
      this.inorder_traversal(node.right, tree);
    }

    return tree;
  }

  count_unival(node)
  {
    let tree = this.inorder_traversal(node, []);
    let count = 0;

    tree.forEach(elem => 
    {
      if(this.is_unival(elem))
      {
        count++;
      }
    });

    return count;
  }

  //TODO: Change this to use dynamic programming instead of a DFS every time.
  is_unival(node)
  {
    if((node === null) || (node.left === null && node.right === null))
    {
      return true;
    }

    else if(node.left === null)
    {
      return (node.data === node.right.data) && (this.is_unival(node.right));
    }

    else if(node.right === null)
    {
      return (node.data === node.left.data) && (this.is_unival(node.left));
    }

    else
    {
      return (node.data === node.left.data) && (node.data === node.right.data) 
              && this.is_unival(node.left) && this.is_unival(node.right);
    }
  }
}

tree = new Node(0, new Node(1), new Node(0, new Node(1, new Node(1), new Node(1)), new Node(0)))
console.log(tree.count_unival(tree)) // 5

tree = new Node(1, new Node(1, new Node(1), new Node(1)), new Node(1, new Node(1), new Node(1)))
console.log(tree.count_unival(tree)) // 7

tree = new Node(0, new Node(1), new Node(2))
console.log(tree.count_unival(tree)) // 2

tree = new Node(0, new Node(1, new Node(2, new Node(3, new Node(4)))))
console.log(tree.count_unival(tree)) // 1

File: Python/DCP08_count_unival.py
class Node:
  def __init__(self, data, left=None, right=None):
    self.data = data
    self.left = left
    self.right = right

  def inorder_traversal(self, node, tree):
    if(node.left):
      self.inorder_traversal(node.left, tree)
    tree.append(node)
    if(node.right):
      self.inorder_traversal(node.right, tree)
    return tree
  
  def count_unival(self, node):
    tree = self.inorder_traversal(node, [])
    count = 0
    for node in tree:
      if self.is_unival(node):
        count = count + 1
    return count

  # TODO: Change this to use dynamic programming instead of a DFS every time.
  def is_unival(self, node):
    if(node is None or (node.left is None and node.right is None)):
      return True
    elif(node.left is None):
      return (node.data is node.right.data) and self.is_unival(node.right)
    elif(node.right is None):
      return (node.data is node.left.data) and self.is_unival(node.left)
    else:
      return (node.data is node.left.data is node.right.data) and self.is_unival(node.left) and self.is_unival(node.right)

tree = Node(0, Node(1), Node(0, Node(1, Node(1), Node(1)), Node(0)))
print(tree.count_unival(tree)) # 5

tree = Node(1, Node(1, Node(1), Node(1)), Node(1, Node(1), Node(1)))
print(tree.count_unival(tree)) # 7

tree = Node(0, Node(1), Node(2))
print(tree.count_unival(tree)) # 2

tree = Node(0, Node(1, Node(2, Node(3, Node(4)))))
print(tree.count_unival(tree)) # 1
